src/parcsr_mv/par_csr_fffc_device.c

```diff
@@ -5,11 +5,179 @@
+   {
+      option = option_;
+      row_CF_marker = row_CF_marker_;
+      col_CF_marker = col_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      if (option == 1)
+      {
+         /* A_{F,F} */
+         return row_CF_marker[i] <   0 && (j == -2 || j >= 0 && col_CF_marker[j] < 0);
+      }
+      else
+      {
+         /* A_{F2, F} */
+         return row_CF_marker[i] == -2 && (j == -2 || j >= 0 && col_CF_marker[j] < 0);
+      }
+   }
+};
+
+/* this predicate selects A^s_{FC} */
+template<typename T>
+struct FC_pred
+{
+   HYPRE_Int *row_CF_marker;
+   T         *col_CF_marker;
+
+   FC_pred(HYPRE_Int *row_CF_marker_, T *col_CF_marker_)
+   {
+      row_CF_marker = row_CF_marker_;
+      col_CF_marker = col_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      return row_CF_marker[i] < 0 && (j >= 0 && col_CF_marker[j] >= 0);
+   }
+};
+
+/* this predicate selects A^s_{CF} */
+template<typename T>
+struct CF_pred
+{
+   HYPRE_Int *row_CF_marker;
+   T         *col_CF_marker;
+
+   CF_pred(HYPRE_Int *row_CF_marker_, T *col_CF_marker_)
+   {
+      row_CF_marker = row_CF_marker_;
+      col_CF_marker = col_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      return row_CF_marker[i] >= 0 && (j >= 0 && col_CF_marker[j] < 0);
+   }
+};
+
+/* this predicate selects A^s_{CC} */
+template<typename T>
+struct CC_pred
+{
+   HYPRE_Int *row_CF_marker;
+   T         *col_CF_marker;
+
+   CC_pred(HYPRE_Int *row_CF_marker_, T *col_CF_marker_)
+   {
+      row_CF_marker = row_CF_marker_;
+      col_CF_marker = col_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      return row_CF_marker[i] >= 0 && (j == -2 || j >= 0 && col_CF_marker[j] >= 0);
+   }
+};
+
+/* this predicate selects A^s_{C,:} */
+struct CX_pred
+{
+   HYPRE_Int *row_CF_marker;
+
+   CX_pred(HYPRE_Int *row_CF_marker_)
+   {
+      row_CF_marker = row_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      return row_CF_marker[i] >= 0 && (j == -2 || j >= 0);
+   }
+};
+
+/* this predicate selects A^s_{:,C} */
+template<typename T>
+struct XC_pred
+{
+   T         *col_CF_marker;
+
+   XC_pred(T *col_CF_marker_)
+   {
+      col_CF_marker = col_CF_marker_;
+   }
+
+   constexpr bool operator()(const Tuple& t) const
+   {
+      const HYPRE_Int i = std::get<0>(t);
+      const HYPRE_Int j = std::get<1>(t);
+
+      return (j == -2 && col_CF_marker[i] >= 0) || (j >= 0 && col_CF_marker[j] >= 0);
+   }
+};
+
```

Comment:
It seems that we can remove some code duplications in the above struct's.

Reply:
Yeah, there's probably a nicer way of doing this... I found myself doing a lot of if/defing due to the use of thrust functionality, inheriting from thrust classes, etc. So I ended up just copy/pasting the whole thing. I'll go through and try unifying it again and see what it looks like and we can decide what we like better.